fix visualization/draw ICP example and add warnings

rxba opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Type

<!--- Select with 'x' and link to a related issue. What types of changes does your code introduce? -->

-   [x] Bug fix (non-breaking change which fixes an issue): Fixes #6887 
-   [ ] New feature (non-breaking change which adds functionality). Resolves #
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) Resolves #

## Motivation and Context

A minor issue, but the ICP example in `examples/python/visualization/draw.py` requires the user to select point correspondences for a preliminary rough transformation before ICP registration, but users may not be aware of this. Currently, if no point correspondences are selected, the application crashes with an IndexError.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that
apply.  If you're unsure about any of these, don't hesitate to ask. We're here
to help! -->

-   [x] I have run `python util/check_style.py --apply` to apply Open3D **code style**
    to my code.
-   [ ] This PR changes Open3D behavior or adds new functionality.
    -   [ ] Both C++ (Doxygen) and Python (Sphinx / Google style) **documentation** is
        updated accordingly.
    -   [ ] I have added or updated C++ and / or Python **unit tests** OR included **test
        results** (e.g. screenshots or numbers) here.
-   [x] I will follow up and update the code if CI fails.
    <!-- In case I am unavailable later -->
-   [x] For fork PRs, I have selected **Allow edits from maintainers**.

## Description
Here are the changes I've made:
* prevented the IndexError when running the ICP example without point correspondences and added a RuntimeWarning for the user to hightlight the need to select point correspondences.
* additionally prevented a KeyError when the "Source (yellow)" cloud is not contained in the first selection set and added a RuntimeWarning for this.
* refactored the existing `do_ICP_one_set` and `do_ICP_two_set` with new functions `_prep_correspondences` and `_do_ICP` to reduce duplicate code.

ssheorey commented on examples/python/visualization/draw.py:
```diff
@@ -142,30 +143,30 @@ def selections():
-        sets = o3dvis.get_selection_sets()
         source_set = sets[0][source_name]
-        target_set = sets[1][target_name]
+        if two_set:
```
Still get an error if the user only selects one set and asks for two_set ICP:

```py
    target_set = sets[1][target_name]
IndexError: list index out of range
```

Can you also add a check for `len(sets)==2`?

ssheorey reviewed (changes requested):
Thanks for this much needed bug fix! Looks good after adding one more check (see below).

rxba commented:
Thank you for the feedback! I've added your suggested check, and also found two more things that needed checking to avoid a crash, which I also added:
* add check if only one set provided to two_set ICP
* add check if no Target (blue) points are provided to single set ICP
* add check if overall number of corresponding points is mismatching and report the mismatch so that the user can add to the existing sets

ssheorey reviewed (approved):
Thanks @rxba ! Looks good.
